Ignore signature overrides

xmo-odoo opened the pull request:
Fixes #46

I included 2 versions of the fix (as well as a minor improvement to the test harness, which incidentally is fairly uncomfortable to run).

Version one is what was discussed in #46, using sphinx's regex as-is, however because the sphinx regex is (apparently) designed to run on trimmed split lines it felt a bit meh, as it requires splitting out the first line, checking the match (on the trimmed first line), then doing the replacement.

The second version uses a modified (and slightly more complicated) version of the Sphinx regex -- I don't think it's a big issue as the regex apparently has not been modified once since it was introduced in 2008 (in sphinx 0.6) -- so it can just do the replacement unconditionally via `re`, no faffing about with conditionals or anything.

peterjc reviewed (changes requested):
Are you on Windows? It looks like ``tests/run_tests.sh`` lost the executable attribute. I can restore that if needs be (once the other things are addressed).

Black would reformat both ``tests/test_cases/explicit_signature.py`` and ``flake8_rst_docstrings.py`` but that's easily fixed.

You've added test cases confirming that the explicit signature gets ignored, which is good.

I would like to see another test with an explicit signature AND an RST violation later in the same docstring - specifically to confirm the violation line number is correct. For example, add an explicit function signature to ``tests/RST203/def_list.py``, or ``tests/RST204/enumerated.py``, or ``tests/RST207/literal.py`` and update the expected output at the start of the file. For now the line number must be checked manually (cross reference #49).

My worry is that when the function line is stripped off the start of the docstring, this leaves the line numbers offset.

peterjc commented:
I don't quite follow how the flake8 stdout change helps with the test suite (I forget now if that was needed on an older version of flake8, or just me not wanting to worry about stdout vs stderr). But yes, the test suite is crude.

Another note to self: Must run this locally and later migrate from TravisCI for continuous integration.

xmo-odoo commented:
Looks like I was fighting with the pre-commit while you reviewed the PR, sorry for the wasted time, those bits should be fixed now:

> Are you on Windows? It looks like tests/run_tests.sh lost the executable attribute. I can restore that if needs be (once the other things are addressed).

Not on windows, I may have screwed up when trying to run it e.g. probably mistook the error I got because of incorrect CWD for a non-runnable, so chmodded +x, didn't work, assumed it had been -x previous, set that, and missed it while reviewing. Should be +x'd again.

> Black would reformat both tests/test_cases/explicit_signature.py and flake8_rst_docstrings.py but that's easily fixed.

Yep, fixed, except part of the new test case as black complaints about something I specifically wanted to test, I turned formatting off for that one bit. I also added the diff so running `pre-commit` locally would provide that info instead of being silent.

> I would like to see another test with an explicit signature AND an RST violation later in the same docstring - specifically to confirm the violation line number is correct. For example, add an explicit function signature to tests/RST203/def_list.py, or tests/RST204/enumerated.py, or tests/RST207/literal.py and update the expected output at the start of the file. For now the line number must be checked manually (cross reference #49).

> My worry is that when the function line is stripped off the start of the docstring, this leaves the line numbers offset.

It *should* be fine because I'm specifically replacing just the signature and leaving the line in-place (just empty) specifically to avoid this issue, but you're right that I forgot to actually check it.

peterjc commented:
Thanks for those quick fixes.

I'm happy with adding ``--diff`` to the black entry in the pre-commit configuration.

However, rather than complicating the pre-commit configuration by skipping the black on one  test case, I would rather just markup the file as needed with a ``# fmt: off`` to ``# fmt: on`` block, or adding ``# fmt: skip`` to a single problematic statement.

Do you have an example where ``re.MULTILINE`` is used? That would probably break the line offsets if a multiline function signature is replaced with a single blank line.

xmo-odoo commented:
> However, rather than complicating the pre-commit configuration by skipping the black on one test case, I would rather just markup the file as needed with a # fmt: off to # fmt: on block, or adding # fmt: skip to a single problematic statement.

That's what I did, the update to the pre-commit is because of flake8, specifically pydocstyle: it *really* does not like signatures as first line, on account of both D400 (wants first line to end with a period) and D402 (specifically forbids signatures).

Obviously that also triggers when adding a sig' to e.g. R203. Maybe I should just disable D400 and D402? That's not great though because it means they're not checked on the package itself.

The alternative might be to add support for *comments*. It looks like Sphinx somehow handles it as part of its preprocessing (at $dayjob the method with an explicit signature on which I used `# noqa: RST210` does render properly through autodoc) but I've no idea where that happens: the extraction of the docstring calls `sphinx.utils.inspect.getdoc` but that only deals with inheritance, and `sphinx.util.docstrings.prepare_docstring` but that has no comments stuff anywhere. All the mentions of "comment" in autodoc *seem* related to annotating via `#:`.

> Do you have an example where `re.MULTILINE` is used? That would probably break the line offsets if a multiline function signature is replaced with a single blank line.

That should not really work since `re.DOTALL` is not enabled, meaning `.` does not match newlines, so only `\s` can, which is limited to some interstitial spaces in the original (betwee tokens of the signature, and between the word and the parameters list, but crucially *nowhere inside teh parameters list* which is where you'd normally expect linebreaks).

The `\s` is what caused the issue though, `\s` does match newlines by default so the "trailing whitespace" ate the first newline. Maybe I should replace the `\s`s by just `[ ]` to ensure only actual spaces are matched? Seems unlikely anyone would want to use a tab in a signature, and `\v` or `\f` are even wonkier.

xmo-odoo commented:
I just checked on our function with `# noqa: RST210` and the signature handling definitely breaks when adding a newline inside the params list, so probably not much of a concern. Still have no idea how sphinx handles the comment yet.

BTW you didn't say whether you preferred option A or option B for stripping the sig (option A would not actually have the whitespace issue since it works on a linewise basis, though it would still have the comments / noqa issue, and it's... less nice I think).

peterjc commented on .pre-commit-config.yaml:
```diff
@@ -29,21 +29,21 @@ repos:
     rev: 4.0.1
     hooks:
     -   id: flake8
+        args: [--exclude, tests/test_cases/explicit_signature.py]
```
Please remove this line, instead special ``# noqa ...`` comments to ``tests/test_cases/explicit_signature.py`` as needed.

peterjc commented:
Sorry, you were excluding files for flake8 (not black) in the pre-commit config. In which case adding ``# noqa ...`` lines to ``tests/test_cases/explicit_signature.py`` is preferable. e.g. https://github.com/peterjc/flake8-rst-docstrings/blob/v0.2.3/tests/RST201/blocks.py#L8

peterjc commented:
Can function signature lines be multi-line? If so that seems problematic for the second approach (B) with a single regex to remove it (leaving a blank line). If not, the history as is seems fine (i.e. Show how the regex was extended going from A to B, which will help if and when it needs to be updated).

peterjc commented on .pre-commit-config.yaml:
```diff
@@ -29,21 +29,21 @@ repos:
     rev: 4.0.1
     hooks:
     -   id: flake8
+        args: [--exclude, tests/test_cases/explicit_signature.py]
```
i.e. for the function docstring with the signature, after the closing triple quote, can you not add ``  # noqa D400,D402``?

peterjc commented:
Personally I don't like overly complicated regular expressions, and the fine details of ``\s`` versus ``\v`` and ``\f`` are all things I would have to look up. Not worrying about tabs and only allowing optional leading/trailing spaces seems reasonable, but leaves a corner case open.

Perhaps it would be wiser to stick with your first approach (using the Sphinx regex unchanged)?

xmo-odoo commented:
> (GitHub seems slow to add comments, which is why I'm only replying to the regex details now - I see it via email)

No worries.

To answer the questions and provide the results of my investigation:

Sphinx does not actually really handle comments, it's just sheer luck that it worked for $dayjob's function: when the function has a return annotation (e.g. `-> foo`) then the comment just gets parsed as part of the type, then most likely whatever sphinx uses to eval the actual type strips out comments.

If the return annotation is not present that doesn't work, so in essence the comments do work no worse here than they do in sphinx (aka they work as as side effect and only in the presence of a return annotation). Therefore I can just use `# noqa` indeed and it *should* work fine as long as there's a return annotation, though I'll have to test that.

That is:
```python
def foo(*args, **kwargs):
    """foo([a][, **others]) -> test # noqa: D400,D402

    Does a thing.
    """
```
```
<sphinx.ext.autodoc.FunctionDocumenter object at 0x7fdb96c3c190> foo([a][, **others]) -> test # noqa: D400,D402              
<re.Match object; span=(0, 46), match='foo([a][, **others]) -> test # noqa: D400,D402'>
```
but
```python
def foo(*args, **kwargs):
    """foo([a][, **others])# noqa: D400,D402

    Does a thing.
    """
```
```
<sphinx.ext.autodoc.FunctionDocumenter object at 0x7fb7038cda50> foo([a][, **others])# noqa: D400,D402                       
None

/tmp/testsig/doc/../src/thing.py:docstring of thing.foo:1: WARNING: Inline strong start-string without end-string.
```
which turns out to be quite funny, if not super understandable while reading the code.

Explicitly handling comments (and thus somewhat formally supporting noqa) could also be an alternative still though.

> Can function signature lines be multi-line?

Nope, the sphinx *regex* looks like they could but somehow I managed to fail to consider that it only ever runs on the first line of the docstring, so as far as sphinx is concerned a signature can not be broken up over multiple lines.

> Personally I don't like overly complicated regular expressions, and the fine details of `\s` versus `\v` and `\f` are all things I would have to look up. Not worrying about tabs and only allowing optional leading/trailing spaces seems reasonable, but leaves a corner case open.

`\s` is an alias for the unicode `White_Space` property, or for the set {space, tab, line feed, form feed, carriage return, vertical tab} in ASCII (aka `[ \t\n\r\f\v]`.

We stop before the first LF/CR, and in my experience `\v` and `\f` are at best errors so I think they matter even less than tabs. So replacing the `\s*` by `[ ]*` (or `\x20*` but I'm not sure that's any clearer, sadly `\N` is only available from 3.8) would probably do what we'd want.

> Perhaps it would be wiser to stick with your first approach (using the Sphinx regex unchanged)?

Maybe. I'm not fond of what it does to the code though, and it has all the issues of unexplained / implicit behaviours of the original.

peterjc commented:
I think this:

```python
def foo(*args, **kwargs):
    """foo([a][, **others]) -> test # noqa: D400,D402

    Does a thing.
    """
```

Should be:

```python
def foo(*args, **kwargs):
    """foo([a][, **others]) -> test

    Does a thing.
    """  # noqa: D400,D402
```

And good to know that function signatures must apparently be one (potentially very long) line. At least with the current version of Sphinx...

xmo-odoo commented:
> I think this:
> 
> ```python
> def foo(*args, **kwargs):
>     """foo([a][, **others]) -> test # noqa: D400,D402
> 
>     Does a thing.
>     """
> ```
> 
> Should be:
> 
> ```python
> def foo(*args, **kwargs):
>     """foo([a][, **others]) -> test
> 
>     Does a thing.
>     """  # noqa: D400,D402
> ```
> 
> And good to know that function signatures must apparently be one (potentially very long) line. At least with the current version of Sphinx...

Oh? Well that makes everything much simpler and basically resolves the issue completely, I didn't know flake treated the entire string as a single line for purpose of warnings.

xmo-odoo commented:
@peterjc added the correct noqa invocations to the new test case, replaced the over-eager `\s` by just looking for spaces, and added a signature to RST203 to verify that the line offsets (on error) are not broken by this change.

peterjc commented:
Rebased after applying most of your other work like adopting pytest.

peterjc reviewed (changes requested):
Having thought about this, I'd prefer the simpler regex from Sphinx autodoc. The Python code might then be made more concise along these lines:

```python
                # ignore the first line if it's a signature
                firstline = docstring.splitlines()[0]
                if py_ext_sig_re.match(firstline.strip()):
                    # replace by an empty line so the line number
                    # doesn't need to be adjusted
                    docstring = docstring.replace(firstline, "")

                try:
                    rst_errors = list(rst_lint.lint(docstring))
```

xmo-odoo commented:
@peterjc sure thing, there you go

peterjc commented:
Refreshed to include black style & your extra test cases.

peterjc commented:
Refreshed *again* to include black style & your extra test cases, *and* the extended commit message.

peterjc commented:
Are you OK with being named in the README file for these recent contributions? See e.g. v0.0.10 and v0.0.9 entries.
